src/uct/ib/base/ib_iface.h

```diff
@@ -304,6 +305,46 @@ static inline uint8_t uct_ib_iface_umr_id(uct_ib_iface_t *iface)
                       _iface, _i, ibv_wc_status_str(_wc[i].status), \
                       &_wc[i], (unsigned long long)_wc[i].wr_id); \
         } \
-               1; }); ++_i) 
+               1; }); ++_i)
+
+/**
+ * Fill ibv_sge data structure by data provided in uct_iov_t
+ * The function avoids copying IOVs with zero length
+ *
+ * @return Number of elements in sge[]
+ */
+static UCS_F_ALWAYS_INLINE
+size_t uct_rc_verbs_rdma_zcopy_sge_fill_iov(struct ibv_sge *sge,
+                                            const uct_iov_t *iov, size_t iovcnt)
+{
+    size_t iov_it = 0, sge_it = 0;
+    uct_ib_mem_t *ib_memh = NULL;
+
+    for (iov_it = 0; iov_it < iovcnt; ++iov_it) {
+        sge[sge_it].length = 0;
+        sge[sge_it].addr   = 0;
+        if (iov[sge_it].count > 1) {
+            sge[sge_it].length = (iov[iov_it].count - 1) * iov[iov_it].stride;
+        }
+        sge[sge_it].length += iov[iov_it].length;
+
+        if (sge[sge_it].length > 0) {
+            sge[sge_it].addr   = (uintptr_t)(iov[iov_it].buffer);
+        } else {
+            continue; /* to avoid zero length elements in sge */
+        }
+
+        if (iov[sge_it].memh == UCT_INVALID_MEM_HANDLE) {
+            sge[sge_it].lkey = 0;
+        } else {
+            ib_memh = (uct_ib_mem_t *)(iov[iov_it].memh);
+            sge[sge_it].lkey = ib_memh->lkey;
```

Comment:
can be one-liner: `sge[sge_it].lkey = ((uct_ib_mem_t *)iov[iov_it].memh)->lkey;`